dali/kernels/slice/slice_cpu_test.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gtest/gtest.h>
+#include "dali/kernels/slice/slice_kernel_test.h"
+#include "dali/kernels/slice/slice_cpu.h"
+
+namespace dali {
+namespace kernels {
+
+template <typename TestArgs>
+class SliceCPUTest : public SliceTest<TestArgs> {
+ public:
+  using InputType = typename TestArgs::InputType;
+  using OutputType = typename TestArgs::OutputType;
+  static constexpr std::size_t Dims = TestArgs::Dims;
+  static constexpr std::size_t NumSamples = TestArgs::NumSamples;
+  static constexpr std::size_t DimSize = TestArgs::DimSize;
+  using SliceArgsGenerator = typename TestArgs::SliceArgsGenerator;
+  using KernelType = SliceCPU<OutputType, InputType, Dims>;
+
+  void Run() override {
+    KernelContext ctx;
+
+    TestTensorList<InputType, Dims> test_data;
+    SliceTest<TestArgs>::PrepareData(test_data);
+
+    auto test_data_cpu = test_data.cpu();
+    auto slice_args = SliceTest<TestArgs>::GenerateSliceArgs(test_data_cpu);
+
+    TestTensorList<OutputType, Dims> expected_output;
+    SliceTest<TestArgs>::PrepareExpectedOutput(test_data, slice_args, expected_output);
+
+    TensorListShape<> output_shapes;
+    output_shapes.resize(NumSamples, Dims);
+    std::vector<KernelType> kernels(NumSamples);
+    for (std::size_t i = 0; i < NumSamples; i++) {
+      auto &kernel = kernels[i];
+      KernelRequirements kernel_req = kernel.Setup(ctx, test_data_cpu[i], slice_args[i]);
```

Comment:
I know that this particular kernel is stateless, but in general case, subsequent calls to `Setup` will overwrite whatever previous calls have done to the kernels internal state. You should have a kernel instance per sample or process on a fully per-sample basis.

Reply:
it is one per sample now, isn't it?

Reply:
Now it is. As the operators go, you can look at Resize, where I used a kernel instance per CPU worker thread.